### PR TITLE
MAT-229: unit tests for C++ OGTerminals

### PR DESCRIPTION
**Pull request by stuartarchibald**
This PR does:
- Refactoring of `expression.hh` to remove terminal types into `terminal.hh` and the `OGNumeric` base class into `numeric.hh`
- Implements sparse matrix `toArrayOfArrays()` method
- Tests the C++ terminal types, ctors, methods etc.

BB PASS: http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/187

### CODE REVIEW

File: src/librdag/test/check_expressions.cc [hunk 1]
```diff
@@ -201,7 +202,7 @@
 }
 
 TEST(VirtualCopyTest, RealSparseMatrix){
-  int colPtr[2] = { 0, 1 };
+  int colPtr[3] = { 0, 2, 2 };
```
Comment (gmarkall) [inline src/librdag/test/check_expressions.cc:205]: I think this changes the matrix from being [ 1.0 0.0; 0.0 2.0 ] to [ 1.0 2.0; 0.0 0.0 ] as well as fixing the length of colPtr. I think this doesn't matter for the purpose of this test though.

**Comment by gmarkall** [comment]
This looks good. I like the FakeVisitor/hasBeenVisited check, it's a good idea.

I'm happy to merge this if you are happy - do you agree that the actual matrix structure mentioned in the comment above is irrelevant, given that the structure is now valid?

File: src/librdag/test/check_expressions.cc (hunk 1 above)
Comment (stuartarchibald) [inline src/librdag/test/check_expressions.cc:205]: t'other way up as its CSC but yeah doesn't matter.

**Comment by stuartarchibald** [comment]
Yep, agree it doesn't matter at all so long as the reads are at valid addresses.

**Pull request merged**